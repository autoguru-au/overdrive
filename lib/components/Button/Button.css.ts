import { style } from '@vanilla-extract/css';
import { recipe, RecipeVariants } from '@vanilla-extract/recipes';

import { focusOutlineStyle } from '../../styles/focusOutline.css';
import { themeContractVars as vars } from '../../themes/theme.css';

const intentColors = vars.colours.intent;
const smallHeight = '36px';

const selectorFocusHoverActive =
	'&:focus-visible, &:not(:disabled):hover, &:not(:disabled):active';

export const body = style({
	display: 'grid',
	gridAutoFlow: 'column dense',
	gridGap: vars.space['1'],
});

export const hiddenContent = style({
	visibility: 'hidden',
});

export const spinner = style({
	margin: '0 auto',
});

// Button recipe with all variants
export const button = recipe({
	base: [
		{
			transitionTimingFunction: vars.animation.easing.standard,
			transitionDuration: '0.1s',
			transitionProperty:
				'color, background-color, border-color, box-shadow, transform',
			transform: 'translate(0, 0) scale(1)',
			willChange: 'transform',
			cursor: 'pointer',
			selectors: {
				'&:active:not(:disabled, [data-loading])': {
					transform: 'scale(0.97)',
				},
				'&[data-loading], &:disabled': {
					cursor: 'not-allowed',
				},
				'&:not([data-loading]):disabled': {
					opacity: '0.3',
				},
			},
		},
		focusOutlineStyle,
	],

	variants: {
		// Size variants
		size: {
			small: {
				height: smallHeight,
			},
			medium: {
				height: vars.space['8'],
			},
			xsmall: {
				padding: `2px ${vars.space['2']}`,
			},
		},
		// Shape variants
		shape: {
			default: {},
			rounded: {},
			iconOnly: {},
		},
		// Intent (color scheme) variants
		intent: {
			primary: {
				color: intentColors.primary.foreground,
				backgroundColor: intentColors.primary.background.standard,
				[selectorFocusHoverActive]: {
					color: intentColors.primary.foreground,
					backgroundColor: intentColors.primary.background.strong,
				},
			},
			brand: {
				color: intentColors.brand.foreground,
				backgroundColor: intentColors.brand.background.standard,
				[selectorFocusHoverActive]: {
					backgroundColor: intentColors.brand.background.strong,
				},
			},
			secondary: {
				color: intentColors.secondary.foreground,
				backgroundColor: intentColors.secondary.background.standard,
				border: `1px solid ${intentColors.secondary.border}`,
				selectors: {
					[selectorFocusHoverActive]: {
						backgroundColor:
							intentColors.secondary.background.strong,
						borderColor: intentColors.secondary.background.strong,
					},
				},
			},
			danger: {
				backgroundColor: intentColors.danger.background.standard,
				color: intentColors.danger.foreground,
				[selectorFocusHoverActive]: {
					backgroundColor: intentColors.danger.background.strong,
				},
			},
			information: {
				backgroundColor: intentColors.information.background.standard,
				color: intentColors.information.foreground,
				[selectorFocusHoverActive]: {
					backgroundColor: intentColors.information.background.strong,
				},
			},
			warning: {
				backgroundColor: intentColors.warning.background.standard,
				color: intentColors.warning.foreground,
				[selectorFocusHoverActive]: {
					backgroundColor: intentColors.warning.background.strong,
				},
			},
			success: {
				backgroundColor: intentColors.success.background.standard,
				color: intentColors.success.foreground,
				[selectorFocusHoverActive]: {
					backgroundColor: intentColors.success.background.strong,
				},
			},
		},
		// Miminal appearance variant
		minimal: {
			true: {
				color: vars.typography.colour.neutral,
				backgroundColor: 'transparent',
				border: 'none',
			},
			false: [],
		},
	},
	compoundVariants: [
		// Size and shape compound variants
		{
			variants: { size: 'small', shape: 'default' },
			style: {
				minWidth: vars.space['8'],
				gridGap: vars.space['1'],
			},
		},
		{
			variants: { size: 'small', shape: 'rounded' },
			style: {
				minWidth: smallHeight,
			},
		},
		{
			variants: { size: 'small', shape: 'iconOnly' },
			style: {
				width: smallHeight,
			},
		},
		{
			variants: { size: 'medium', shape: 'default' },
			style: {
				minWidth: vars.space['9'],
				gridGap: vars.space['2'],
			},
		},
		{
			variants: { size: 'medium', shape: 'rounded' },
			style: {
				minWidth: vars.space['8'],
			},
		},
		{
			variants: { size: 'medium', shape: 'iconOnly' },
			style: {
				width: vars.space['8'],
			},
		},
		// Minimal compound variants per intent
		{
			variants: { intent: 'primary', minimal: true },
			style: {
				selectors: {
					[selectorFocusHoverActive]: {
						color: intentColors.primary.background.strong,
						backgroundColor: intentColors.primary.background.mild,
					},
				},
			},
		},
		{
			variants: { intent: 'brand', minimal: true },
			style: {
				selectors: {
					[selectorFocusHoverActive]: {
						color: intentColors.brand.background.strong,
						backgroundColor: intentColors.brand.background.mild,
					},
				},
			},
		},
		{
			variants: { intent: 'secondary', minimal: true },
			style: {
				selectors: {
					[selectorFocusHoverActive]: {
						color: vars.typography.colour.secondary,
						backgroundColor:
							intentColors.secondary.background.strong,
					},
				},
			},
		},
		{
			variants: { intent: 'danger', minimal: true },
			style: {
				selectors: {
					[selectorFocusHoverActive]: {
						color: intentColors.danger.background.strong,
						backgroundColor: intentColors.danger.background.mild,
					},
				},
			},
		},
		{
			variants: { intent: 'information', minimal: true },
			style: {
				selectors: {
					[selectorFocusHoverActive]: {
						color: intentColors.information.background.strong,
						backgroundColor:
							intentColors.information.background.mild,
					},
				},
			},
		},
		{
			variants: { intent: 'warning', minimal: true },
			style: {
				selectors: {
					'&:focus-visible, &:not(:disabled):hover': {
						color: intentColors.warning.background.strong,
						backgroundColor: intentColors.warning.background.mild,
					},
				},
			},
		},
		{
			variants: { intent: 'success', minimal: true },
			style: {
				selectors: {
					'&:focus-visible, &:not(:disabled):hover': {
						color: intentColors.success.background.strong,
						backgroundColor: intentColors.success.background.mild,
					},
				},
			},
		},
	],
	defaultVariants: {
		size: 'medium',
		shape: 'default',
		intent: 'primary',
		minimal: false,
	},
});

type ButtonRecipeProps = NonNullable<Required<RecipeVariants<typeof button>>>;

export type ButtonSize = ButtonRecipeProps['size'];
export type ButtonShape = ButtonRecipeProps['shape'];
export type ButtonIntent = ButtonRecipeProps['intent'];
export type ButtonMinimal = ButtonRecipeProps['minimal'];

export interface StyledButtonProps {
	/**
	 * Button sizing
	 */
	size?: ButtonSize;
	/**
	 * Button intentional colour scheme
	 */
	variant?: ButtonIntent;
	/**
	 * Present a borderless minimal appearance
	 */
	minimal?: ButtonMinimal;
}
