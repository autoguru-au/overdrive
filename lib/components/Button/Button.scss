@import '../../theme/tokens/facade';

@mixin isState($state, $nest: false) {
	@if $nest {
		$sel: &;

		@at-root .root:not([disabled]):#{$state} {
			#{$sel} {
				@content;
			}
		}
	} @else {
		@at-root &.root:not([disabled]):#{$state} {
			@content;
		}
	}
}

.root {
	display: inline-block;
	overflow: hidden;
	width: auto;
	outline: none;
	font-family: inherit;
	line-height: 1;
	font-weight: 500;
	text-align: center;
	color: white;
	fill: currentColor;
	background: none;
	border: none;
	text-decoration: none;
	cursor: pointer;
	user-select: none;
	transform: translate(0, 0) scale(1);
	will-change: transform;
	border-radius: var(--global--space--1);
	transition-delay: 0s;
	transition-duration: 0.2s;
	transition-timing-function: cubic-bezier(0, 0, 0.2, 1);
	transition-property: color, background-color, border-color, box-shadow,
		transform;

	&[disabled] {
		cursor: not-allowed;

		&:not(.loading) {
			opacity: 0.3;
		}
	}

	@include isState('active') {
		&:not(.minimal) {
			transform: translate(0, 1px) scale(1);
		}
	}
}

.body {
	display: grid;
	grid-auto-flow: column dense;
	grid-gap: var(--global--space--1);
	justify-content: center;
	align-items: center;
	height: 100%;
}

// Sizes
.small,
.medium {
	&.minimal {
		min-width: auto;
	}
}

.small {
	$height: 36px;

	grid-gap: var(--global--space--1);
	font-size: var(--global--fontSize--3);
	line-height: var(--global--fontSize--3--lh);
	min-width: 48px;
	height: $height;

	&:not(.loading) {
		padding: 0 var(--global--space--3);
	}

	&.rounded {
		min-width: $height;
		border-radius: calc(#{$height} / 2);
	}

	&.iconOnly {
		width: 36px;
		height: 36px;
	}
}

.medium {
	$height: 48px;

	grid-gap: var(--global--space--2);
	font-size: var(--global--fontSize--4);
	line-height: var(--global--fontSize--4--lh);
	min-width: 96px;
	height: $height;

	&:not(.loading) {
		padding: 0 var(--global--space--4);
	}

	&.rounded {
		min-width: $height;
		border-radius: calc(#{$height} / 2);
	}

	&.iconOnly {
		width: 48px;
		height: 48px;
	}
}

.fullWidth {
	display: grid;
	width: 100%;
}

// Variants
.variantPrimary {
	background-color: getColour(green, 600);
	box-shadow: inset 0 0 0 1px getColour(green, 700),
		0 1px 10px 0 rgba(0, 0, 0, 0.06), 0 4px 5px 0 rgba(0, 0, 0, 0.06),
		0 2px 4px -1px rgba(0, 0, 0, 0.1);

	&.minimal {
		color: getColour(gray, 700);
		background-color: transparent;
		box-shadow: none;
	}

	@include isState('hover') {
		color: white;
		background-color: getColour(green, 700);
		box-shadow: inset 0 0 0 1px getColour(green, 800),
			0 3px 14px 2px rgba(0, 0, 0, 0.06),
			0 8px 10px 1px rgba(0, 0, 0, 0.06),
			0 5px 5px -3px rgba(0, 0, 0, 0.1);

		&.minimal {
			color: getColour(green, 900);
			background-color: getColour(green, 100);
			box-shadow: none;
		}
	}

	@include isState('active') {
		color: white;
		background-color: getColour(green, 800);
		box-shadow: inset 0 0 0 1px getColour(green, 9800),
			0 1px 5px 0 rgba(0, 0, 0, 0.06), 0 2px 2px 0 rgba(0, 0, 0, 0.06),
			0 3px 1px -2px rgba(0, 0, 0, 0.1);

		&.minimal {
			color: getColour(green, 900);
			background-color: getColour(green, 200);
			box-shadow: none;
		}
	}
}

.variantSecondary {
	color: getColour(gray, 700);
	background-color: white;
	box-shadow: inset 0 0 0 1px getColour(gray, 300),
		var(--global--elevationBoxShadow--2);

	&.minimal {
		color: getColour(gray, 700);
		background-color: transparent;
		box-shadow: none;
	}

	@include isState('hover') {
		color: getColour(gray, 700);
		background-color: getColour(gray, 200);
		box-shadow: inset 0 0 0 1px getColour(gray, 400),
			var(--global--elevationBoxShadow--3);

		&.minimal {
			color: getColour(gray, 900);
			background-color: getColour(gray, 100);
			box-shadow: none;
		}
	}

	@include isState('active') {
		color: getColour(gray, 700);
		background-color: getColour(gray, 300);
		box-shadow: inset 0 0 0 1px getColour(gray, 500),
			var(--global--elevationBoxShadow--1);

		&.minimal {
			color: getColour(gray, 900);
			background-color: getColour(gray, 200);
			box-shadow: none;
		}
	}
}

.variantDanger {
	background-color: getColour(red, 600);
	box-shadow: inset 0 0 0 1px getColour(red, 700),
		0 1px 10px 0 rgba(0, 0, 0, 0.09), 0 4px 5px 0 rgba(0, 0, 0, 0.09),
		0 2px 4px -1px rgba(0, 0, 0, 0.15);

	&.minimal {
		color: getColour(gray, 700);
		background-color: transparent;
		box-shadow: none;
	}

	@include isState('hover') {
		color: white;
		background-color: getColour(red, 700);
		box-shadow: inset 0 0 0 1px getColour(red, 800),
			0 3px 14px 2px rgba(0, 0, 0, 0.09),
			0 8px 10px 1px rgba(0, 0, 0, 0.09),
			0 5px 5px -3px rgba(0, 0, 0, 0.15);

		&.minimal {
			color: getColour(red, 900);
			background-color: getColour(red, 100);
			box-shadow: none;
		}
	}

	@include isState('active') {
		color: white;
		background-color: getColour(red, 800);
		box-shadow: inset 0 0 0 1px getColour(red, 900),
			0 1px 5px 0 rgba(0, 0, 0, 0.09), 0 2px 2px 0 rgba(0, 0, 0, 0.09),
			0 3px 1px -2px rgba(0, 0, 0, 0.15);

		&.minimal {
			color: getColour(red, 900);
			background-color: getColour(red, 200);
			box-shadow: none;
		}
	}
}
