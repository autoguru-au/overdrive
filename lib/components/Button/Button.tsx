import { IconType } from '@autoguru/icons';
import clsx from 'clsx';
import React, {
	AriaAttributes,
	ButtonHTMLAttributes,
	cloneElement,
	ComponentProps,
	createElement,
	forwardRef,
	isValidElement,
	ReactElement,
	useMemo,
} from 'react';

import { Icon } from '../Icon';
import { ProgressSpinner } from '../ProgressSpinner';
import styles from './Button.scss';

export enum EButtonSize {
	Medium = 'medium',
	Small = 'small',
}

export enum EButtonVariant {
	Primary = 'primary',
	Secondary = 'secondary',
	Danger = 'danger',
}

type ButtonPrimitive = ButtonHTMLAttributes<HTMLButtonElement>;

type AllowedChildren = string | IconType;

interface Props
	extends Pick<ButtonPrimitive, 'id' | 'onClick' | 'type' | 'className'>,
		Pick<AriaAttributes, 'aria-label'> {
	children: AllowedChildren | AllowedChildren[];
	disabled?: boolean;
	is?: ReactElement | 'button';
	isLoading?: boolean;
	isFullWidth?: boolean;
	minimal?: boolean;
	rounded?: boolean;
	size?: EButtonSize;
	variant?: EButtonVariant;
}

export const Button = forwardRef<HTMLButtonElement, Props>(
	(
		{
			children,
			className = '',
			disabled = false,
			id,
			is: Component = 'button',
			isLoading = false,
			isFullWidth = false,
			minimal = false,
			onClick,
			rounded = false,
			size = EButtonSize.Medium,
			type = 'button',
			variant = EButtonVariant.Secondary,
			'aria-label': ariaLabel,
		},
		ref,
	) => {
		const { isSingleIconChild, props: maybeIconProps } = useMemo(() => {
			const maybeIcon =
				isValidElement(children) && children.type === Icon;
			const maybeProps = children as ReactElement<
				ComponentProps<typeof Icon>
			>;

			return maybeIcon
				? {
						isSingleIconChild: true,
						props: maybeProps.props,
				  }
				: { isSingleIconChild: false };
		}, [children]);

		const props: Partial<ButtonPrimitive> & { ref: typeof ref } = {
			type: Component === 'button' ? (type as any) : void 0,
			id,
			onClick,
			disabled: disabled || isLoading,
			'aria-label': ariaLabel,
			className: clsx(
				styles.root,
				buttonVariantStyleMap[variant],
				buttonSizeStyleMap[size],
				{
					[styles.loading]: isLoading,
					[styles.fullWidth]: isFullWidth,
					[styles.minimal]: minimal,
					[styles.rounded]: rounded,
					[styles.iconOnly]: isSingleIconChild && !isLoading,
				},
				className,
			),
			ref,
		};

		const child = isLoading ? (
			<ProgressSpinner variant={null} />
		) : (
			<div className={styles.body}>
				{isSingleIconChild ? (
					<Icon
						icon={maybeIconProps.icon}
						size={maybeIconProps.size ?? 20}
						{...maybeIconProps}
					/>
				) : (
					children
				)}
			</div>
		);

		return isValidElement(Component)
			? cloneElement(Component, props, child)
			: createElement(Component, props, child);
	},
);

const buttonVariantStyleMap: Record<EButtonVariant, string> = {
	[EButtonVariant.Danger]: styles.variantDanger,
	[EButtonVariant.Primary]: styles.variantPrimary,
	[EButtonVariant.Secondary]: styles.variantSecondary,
};

const buttonSizeStyleMap: Record<EButtonSize, string> = {
	[EButtonSize.Medium]: styles.medium,
	[EButtonSize.Small]: styles.small,
};
