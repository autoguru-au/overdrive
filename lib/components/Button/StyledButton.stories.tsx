import type { Meta, StoryObj } from '@storybook/react-vite';

import { StyledButton } from './StyledButton';

const meta = {
	title: 'Primatives/Buttons/Styled as Button',
	tags: ['polymorphic'],
	component: StyledButton,
	args: {
		as: 'a',
		href: '#link',
		children: 'Button as link',
		variant: 'primary',
		size: 'small',
		minimal: false,
		rounded: false,
		isFullWidth: false,
	},
} satisfies Meta<typeof StyledButton>;

export default meta;
type Story = StoryObj<typeof StyledButton>;

export const ButtonAsLink: Story = {};
