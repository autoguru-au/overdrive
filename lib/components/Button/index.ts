export {
	Button,
	ESize as EButtonSize,
	EVariant as EButtonVariant,
	Props as IButtonProps,
} from './Button';
