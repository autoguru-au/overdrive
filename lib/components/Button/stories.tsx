import { action } from '@storybook/addon-actions';
import { boolean, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import { Button, EButtonSize, EButtonVariant } from '.';
import styles from './style.scss';

const nonRoundedProps = () => ({
	isFullWidth: boolean('Full Width', false),
});

const baseProps = () => ({
	size: select('Size', EButtonSize, EButtonSize.Medium),
	variant: select('Variant', EButtonVariant, EButtonVariant.Primary),
	onClick(e) {
		e.preventDefault();

		return action('Clicked')(e);
	},
});

storiesOf('Components|Buttons', module)
	.add('default', () => (
		<Button
			{...baseProps()}
			{...nonRoundedProps()}
			isLoading={boolean('isLoading', false)}
			disabled={boolean('Disabled', false)}>
			Hello World
		</Button>
	))
	.add('hoverState', () => (
		<Button
			className={styles.isHoverState}
			{...baseProps()}
			{...nonRoundedProps()}>
			Hello World
		</Button>
	))
	.add('activeState', () => (
		<Button
			className={styles.isActiveState}
			{...baseProps()}
			{...nonRoundedProps()}>
			Hello World
		</Button>
	))
	.add('disabledState', () => (
		<Button {...baseProps()} {...nonRoundedProps()} disabled={true}>
			Hello World
		</Button>
	))
	.add('loadingState', () => (
		<Button {...baseProps()} {...nonRoundedProps()} isLoading={true}>
			Hello World
		</Button>
	))
	.add('asAnchor', () => (
		<Button
			href="https://www.autoguru.com.au"
			target="_blank"
			component="a"
			{...baseProps()}
			{...nonRoundedProps()}>
			Hello World
		</Button>
	))
	.add('rounded', () => (
		<Button rounded={true} {...baseProps()}>
			2
		</Button>
	));
