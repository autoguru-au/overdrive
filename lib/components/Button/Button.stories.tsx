import { AccountBoxIcon, ArrowLeftIcon } from '@autoguru/icons';
import type { Meta, StoryObj } from '@storybook/react-vite';
import React, { type ComponentProps } from 'react';
import { fn } from 'storybook/test';

import { Box } from '../Box/Box';
import { Icon } from '../Icon/Icon';
import { Inline } from '../Inline/Inline';

import { Button, type ButtonProps } from './Button';

const meta: Meta<typeof Button> = {
	title: 'Primitives/Buttons',
	tags: ['polymorphic'],
	component: Button,
	args: {
		variant: 'secondary',
		size: 'medium',
		minimal: false,
		rounded: false,
		children: undefined,
		isFullWidth: false,
		isLoading: false,
		disabled: false,
		as: undefined,
		onClick: fn(),
		withDoubleClicks: false,
		testId: 'storybook-button',
	},
	argTypes: {
		as: {
			options: ['button', 'a'],
		},
	},
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Standard: Story = {
	args: {
		children: 'Login',
	},
};

/**
 * Example of a the extra small rounded button with a back arrow icon
 */
export const ExtraSmall: Story = {
	args: {
		children: (
			<>
				<Icon icon={ArrowLeftIcon} />
				Change car
			</>
		),
		size: 'xsmall',
		rounded: true,
	},
};

const TemplateMulti = ({
	children,
	onClick,
	variant,
}: ComponentProps<typeof Button>) => (
	<>
		{['medium', 'small'].map((size) => {
			const args = {
				children,
				onClick,
				size: size as ButtonProps['size'],
				variant,
			};

			return (
				<>
					<Box
						textAlign="right"
						style={{ textTransform: 'capitalize' }}
					>
						{size}
					</Box>
					<Inline space="3">
						<Button {...args}>Login</Button>
						<Button {...args}>
							<Icon icon={AccountBoxIcon} />
							Login
						</Button>

						<Button {...args} aria-label="login">
							<Icon icon={AccountBoxIcon} />
						</Button>

						<Button rounded {...args} aria-label="login">
							<Icon icon={AccountBoxIcon} />
						</Button>

						<Button isLoading {...args}>
							A very very very long button Label
						</Button>

						<Button disabled {...args}>
							Login
						</Button>
					</Inline>
					<div>
						<Button isFullWidth {...args}>
							Full Width
						</Button>
					</div>
					<Inline space="3">
						<Button minimal {...args}>
							Login
						</Button>

						<Button minimal {...args}>
							<Icon icon={AccountBoxIcon} />
							Login
						</Button>

						<Button minimal {...args} aria-label="login">
							<Icon icon={AccountBoxIcon} />
						</Button>

						<Button minimal rounded {...args}>
							1
						</Button>

						<Button minimal rounded {...args} aria-label="login">
							<Icon icon={AccountBoxIcon} />1
						</Button>
						<Button minimal rounded {...args} aria-label="login">
							<Icon icon={AccountBoxIcon} />
						</Button>

						<Button minimal isLoading {...args}>
							Login
						</Button>

						<Button minimal disabled {...args}>
							Login
						</Button>
					</Inline>
				</>
			);
		})}
	</>
);

export const PrimarySet: Story = {
	decorators: [
		(Story) => (
			<div
				style={{
					display: 'grid',
					gridGap: '12px',
					gridAutoFlow: 'row dense',
				}}
			>
				<Story />
			</div>
		),
	],
	args: {
		variant: 'primary',
	},
	render: TemplateMulti,
};

export const SecondarySet: Story = {
	decorators: PrimarySet.decorators,
	args: {
		variant: 'secondary',
	},
	render: TemplateMulti,
};

export const InformationSet: Story = {
	decorators: PrimarySet.decorators,
	args: {
		variant: 'information',
	},
	render: TemplateMulti,
};

export const WarningSet: Story = {
	decorators: PrimarySet.decorators,
	args: {
		variant: 'warning',
	},
	render: TemplateMulti,
};

export const SuccessSet: Story = {
	decorators: PrimarySet.decorators,
	args: {
		variant: 'success',
	},
	render: TemplateMulti,
};
