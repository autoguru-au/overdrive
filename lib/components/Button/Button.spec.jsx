import React from 'react';
import { mount, shallow } from 'enzyme';
import { Button } from './Button';
import { EButtonSize, EButtonVariant } from './index';

describe('<Button />', () => {
	it('should not throw', () =>
		expect(() => shallow(<Button />)).not.toThrow());

	it('should match snapshot for default button', () => {
		expect(shallow(<Button />)).toMatchSnapshot();
	});

	it('should pass on className to dom element', () => {
		const button = mount(<Button className="button-class" />);
		expect(button.hasClass('button-class')).toBeTruthy();
	});

	describe('when as button', () => {
		it('should use html button element by default', () => {
			const button = shallow(<Button />);
			expect(button.type()).toEqual('button');
		});
	});

	describe('when custom component', () => {
		it('should use html anchor element if href value exists', () => {
			const button = shallow(<Button is="a" href="abc" />);
			expect(button.type()).toEqual('a');
		});

		it('should pass href value to the DOM element', () => {
			const button = mount(<Button is="a" href="/abcd" />);
			expect(button.html()).toContain(`href="/abcd"`);
		});

		describe('deprecated still work', () => {
			it('should use html anchor element if href value exists', () => {
				const button = shallow(<Button component="a" href="abc" />);
				expect(button.type()).toEqual('a');
			});

			it('should pass href value to the DOM element', () => {
				const button = mount(<Button component="a" href="/abcd" />);
				expect(button.html()).toContain(`href="/abcd"`);
			});
		});
	});

	describe('when colour variants', () => {
		it('should match snapshot for Primary button', () => {
			expect(
				shallow(<Button variant={EButtonVariant.Primary} />)
			).toMatchSnapshot();
		});

		it('should apply Primary variant styles', () => {
			const button = shallow(<Button variant={EButtonVariant.Primary} />);
			expect(button.hasClass('variantPrimary')).toBeTruthy();
		});

		it('should match snapshot for Secondary button', () => {
			expect(
				shallow(<Button variant={EButtonVariant.Secondary} />)
			).toMatchSnapshot();
		});

		it('should apply Secondary variant styles', () => {
			const button = shallow(
				<Button variant={EButtonVariant.Secondary} />
			);
			expect(button.hasClass('variantSecondary')).toBeTruthy();
		});

		it('should match snapshot for Danger button', () => {
			expect(
				shallow(<Button variant={EButtonVariant.Danger} />)
			).toMatchSnapshot();
		});

		it('should apply Danger variant styles', () => {
			const button = shallow(<Button variant={EButtonVariant.Danger} />);
			expect(button.hasClass('variantDanger')).toBeTruthy();
		});
	});

	describe('when size variant', () => {
		it('should match snapshot for small button', () => {
			expect(
				shallow(<Button size={EButtonSize.Small} />)
			).toMatchSnapshot();
		});

		it('should apply small size styles', () => {
			const button = shallow(<Button size={EButtonSize.Small} />);
			expect(button.hasClass('sizeSmall')).toBeTruthy();
		});

		it('should match snapshot for medium button', () => {
			expect(
				shallow(<Button size={EButtonSize.Medium} />)
			).toMatchSnapshot();
		});

		it('should apply medium size styles', () => {
			const button = shallow(<Button size={EButtonSize.Medium} />);
			expect(button.hasClass('sizeMedium')).toBeTruthy();
		});
	});

	describe('when rounded', () => {
		it('should match snapshot', () => {
			expect(shallow(<Button rounded={true} />)).toMatchSnapshot();
		});

		it('should not apply rounded styles to default button', () => {
			const button = shallow(<Button />);
			expect(button.hasClass('rounded')).toBeFalsy();
		});

		it('should apply rounded styles', () => {
			const button = shallow(<Button rounded={true} />);
			expect(button.hasClass('rounded')).toBeTruthy();
		});
	});

	describe('when isFullWidth', () => {
		it('should match snapshot', () => {
			expect(shallow(<Button isFullWidth={true} />)).toMatchSnapshot();
		});

		it('should apply styles', () => {
			const button = shallow(<Button isFullWidth={true} />);
			expect(button.hasClass('fullWidth')).toBeTruthy();
		});
	});

	describe('when loading', () => {
		it('should match snapshot for default button', () => {
			expect(shallow(<Button isLoading={true} />)).toMatchSnapshot();
		});

		it('should match snapshot for small button', () => {
			expect(
				shallow(<Button size={EButtonSize.Small} isLoading={true} />)
			).toMatchSnapshot();
		});

		it('should match snapshot for medium button', () => {
			expect(
				shallow(<Button size={EButtonSize.Medium} isLoading={true} />)
			).toMatchSnapshot();
		});

		it('should not apply loading styles to default button', () => {
			const button = shallow(<Button />);
			expect(button.hasClass('loading')).toBeFalsy();
		});

		it('should apply loading styles', () => {
			const button = shallow(<Button isLoading={true} />);
			expect(button.hasClass('loading')).toBeTruthy();
		});
	});
});
