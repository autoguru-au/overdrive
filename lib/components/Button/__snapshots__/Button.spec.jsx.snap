// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot for default button 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when as button should use html button element by default 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Danger button 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantDanger"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Primary button 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantPrimary"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when colour variants should match snapshot for Secondary button 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when custom component should use html anchor element if href value exists 1`] = `
<a
  aria-disabled="false"
  class="root sizeMedium variantSecondary"
  href="abc"
>
  <span
    class="btnContent"
  />
</a>
`;

exports[`<Button /> when isFullWidth should match snapshot 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary fullWidth"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when loading should match snapshot for default button 1`] = `
<button
  aria-disabled="true"
  class="root sizeMedium variantSecondary loading"
  disabled=""
>
  <span
    class="btnContent"
  />
  <div
    class="root sizeMedium variantSecondary spinner"
  >
    <svg
      class="circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when loading should match snapshot for medium button 1`] = `
<button
  aria-disabled="true"
  class="root sizeMedium variantSecondary loading"
  disabled=""
>
  <span
    class="btnContent"
  />
  <div
    class="root sizeMedium variantSecondary spinner"
  >
    <svg
      class="circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when loading should match snapshot for small button 1`] = `
<button
  aria-disabled="true"
  class="root sizeSmall variantSecondary loading"
  disabled=""
>
  <span
    class="btnContent"
  />
  <div
    class="root sizeSmall variantSecondary spinner"
  >
    <svg
      class="circular"
      viewBox="25 25 50 50"
    >
      <circle
        class="path"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
      />
    </svg>
  </div>
</button>
`;

exports[`<Button /> when rounded should match snapshot 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary rounded"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when size variant should match snapshot for medium button 1`] = `
<button
  aria-disabled="false"
  class="root sizeMedium variantSecondary"
>
  <span
    class="btnContent"
  />
</button>
`;

exports[`<Button /> when size variant should match snapshot for small button 1`] = `
<button
  aria-disabled="false"
  class="root sizeSmall variantSecondary"
>
  <span
    class="btnContent"
  />
</button>
`;
